extractor/src/main/java/org/schabi/newpipe/extractor/services/bandcamp/extractors/BandcampSearchExtractor.java

```diff
@@ -0,0 +1,139 @@
+import org.schabi.newpipe.extractor.search.InfoItemsSearchCollector;
+import org.schabi.newpipe.extractor.search.SearchExtractor;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+
+public class BandcampSearchExtractor extends SearchExtractor {
+
+    public BandcampSearchExtractor(StreamingService service, SearchQueryHandler linkHandler) {
+        super(service, linkHandler);
+    }
+
+    @Override
+    public String getSearchSuggestion() {
+        return null;
+    }
+
+    @Override
+    public InfoItemsPage<InfoItem> getPage(String pageUrl) throws IOException, ExtractionException {
+        // okay apparently this is where we DOWNLOAD the page and then COMMIT its ENTRIES to an INFOITEMPAGE
+        String html = getDownloader().get(pageUrl).responseBody();
+
+        InfoItemsSearchCollector collector = new InfoItemsSearchCollector(getServiceId());
+
+
+        Document d = Jsoup.parse(html);
+
+        Elements searchResultsElements = d.getElementsByClass("searchresult");
+
+        for (Element searchResult :
+                searchResultsElements) {
+
+            Element resultInfo = searchResult.getElementsByClass("result-info").first();
+
+            String type = resultInfo
+                    .getElementsByClass("itemtype").first().text();
+
+            String image = null;
+            Element img = searchResult.getElementsByClass("art").first()
+                    .getElementsByTag("img").first();
+            if (img != null) {
+                image = img.attr("src");
+            }
+
+            String heading = resultInfo.getElementsByClass("heading").text();
+
+            String subhead = resultInfo.getElementsByClass("subhead").text();
+
+            String url = resultInfo.getElementsByClass("itemurl").text();
+
+            switch (type) {
+                default:
+                    continue;
+                case "FAN":
+                    //collector.commit Channel (?) with heading, url, image
+                    break;
+
+                case "ARTIST":
+                    collector.commit(new BandcampChannelInfoItemExtractor(heading, url, image, subhead));
+                    break;
+
+                case "ALBUM":
+                    String artist = subhead.split(" by")[0];
+                    String length = resultInfo.getElementsByClass("length").text();
+                    int tracks = Integer.parseInt(length.split(" track")[0]);
+                    collector.commit(new BandcampPlaylistInfoItemExtractor(heading, artist, url, image, tracks));
+                    break;
+
+                case "TRACK":
+                    String[] splitBy = subhead.split(" by");
+                    String artist1 = null;
+                    if (splitBy.length > 1) {
+                        artist1 = subhead.split(" by")[1];
+                    }
+                    collector.commit(new BandcampStreamInfoItemExtractor(heading, url, image, artist1));
+                    break;
+            }
```

Comment:
You shouldn't do a lot of stuff here, but in the appropriate `InfoItemExtractor`s as far as that's possible. This code here should be reduced to something like:

```java
            final Element resultInfo = searchResult.getElementsByClass("result-info").first();

            final String type = resultInfo.getElementsByClass("itemtype").first().text();
            
            switch (type) {
                default:
                    continue;
                case "FAN":
                    //collector.commit Channel (?) with heading, url, image
                    break;
                case "ARTIST":
                    collector.commit(new BandcampChannelInfoItemExtractor(resultInfo));
                    break;
                case "ALBUM":
                    collector.commit(new BandcampPlaylistInfoItemExtractor(resultInfo));
                    break;
                case "TRACK":
                    collector.commit(new BandcampStreamInfoItemExtractor(resultInfo));
                    break;
            }
```

Reply:
Like this? 2139c58dbbb4b4c830068b638337688051653399

I don't think this improves the code quality as the code is now duplicate.

Reply:
Not like that. It seems like you get the data in significantly different ways sometimes. I'll take a look soon.

Reply:
@fynngodau: The idea is that you'll have a class like the following:

```java
package org.schabi.newpipe.extractor.services.bandcamp.extractors;

import org.jsoup.nodes.Element;
import org.schabi.newpipe.extractor.channel.ChannelInfoItemExtractor;
import org.schabi.newpipe.extractor.exceptions.ParsingException;

public class BandcampChannelInfoItemExtractor implements ChannelInfoItemExtractor {
    private final Element searchResult;
    private final Element resultInfo;

    public BandcampChannelInfoItemExtractor(final Element searchResult) {
        this.searchResult = searchResult;
        resultInfo = searchResult.getElementsByClass("result-info").first();
    }

    @Override
    public String getName() throws ParsingException {
        return resultInfo.getElementsByClass("heading").text();
    }

    @Override
    public String getUrl() throws ParsingException {
        return resultInfo.getElementsByClass("itemurl").text();
    }

    @Override
    public String getThumbnailUrl() throws ParsingException {
        return searchResult.getElementsByClass("art").first().getElementsByTag("img").first();
    }

    @Override
    public String getDescription() {
        return resultInfo.getElementsByClass("subhead").text();
    }

    @Override
    public long getSubscriberCount() {
        return -1;
    }

    @Override
    public long getStreamCount() {
        return -1;
    }
}
```

Reply:
@wb9688: I changed the `BandcampChannelInfoItemExtractor` as you suggested, as well as the other `InfoItem` classes, in 692b2d06f48d469f965b74093c14060a77f6c555. I hope it is what you had in mind.